contracts/governance/InterchainMultisig.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { IInterchainMultisig } from '../interfaces/IInterchainMultisig.sol';
+import { SafeNativeTransfer } from '../libs/SafeNativeTransfer.sol';
+import { ECDSA } from '../libs/ECDSA.sol';
+import { Caller } from '../utils/Caller.sol';
+import { BaseWeightedMultisig } from './BaseWeightedMultisig.sol';
+
+/**
+ * @title InterchainMultisig Contract
+ * @notice Weighted Multisig executor to call functions on any contract
+ */
+contract InterchainMultisig is Caller, BaseWeightedMultisig, IInterchainMultisig {
+    // keccak256('InterchainMultisig.Storage')
+    bytes32 private constant INTERCHAIN_MULTISIG_STORAGE =
+        0x5a9dc2248a56f285e6221da431581a5990380ebfa07727563571b2be1517a23e;
+
+    using SafeNativeTransfer for address;
+
+    bytes32 public immutable chainNameHash;
+
+    struct InterchainMultisigStorage {
+        mapping(bytes32 => bool) isPayloadExecuted;
+    }
+
+    /**
+     * @notice Contract constructor
+     * @dev Sets the initial list of signers and corresponding threshold.
+     * @param chainName The name of the chain
+     * @param weightedSigners The weighted signers payload
+     */
+    constructor(string memory chainName, WeightedSigners memory weightedSigners) BaseWeightedMultisig(0) {
+        chainNameHash = keccak256(bytes(chainName));
+
+        _rotateSigners(weightedSigners);
+    }
+
+    modifier onlySelf() {
+        if (msg.sender != address(this)) revert NotSelf();
+
+        _;
+    }
+
+    /**
+     * @notice Checks if a payload has been executed
+     * @param payloadHash The hash of the payload payload
+     * @return True if the payload has been executed
+     */
+    function isPayloadExecuted(bytes32 payloadHash) external view returns (bool) {
+        return _interchainMultisigStorage().isPayloadExecuted[payloadHash];
+    }
+
+    /**
+     * @notice Executes an external contract call.
+     * @notice This function is protected by the onlySigners requirement.
+     * @dev Calls a target address with specified calldata and passing provided native value.
+     * @param callBatch The batch of calls to execute
+     * @param proof The multisig proof data
+     */
+    function executeCalls(bytes calldata callBatch, bytes calldata proof) external payable {
+        bytes32 messageHash = ECDSA.toEthSignedMessageHash(keccak256(callBatch));
+        validateProof(messageHash, proof);
+
+        InterchainMultisigStorage storage slot = _interchainMultisigStorage();
+        (bytes32 salt, Call[] memory calls) = abi.decode(callBatch, (bytes32, Call[]));
+        uint256 length = calls.length;
+
+        emit BatchExecuted(messageHash, salt, length);
+
+        if (slot.isPayloadExecuted[messageHash]) revert AlreadyExecuted();
```

Comment:
```suggestion
        if (slot.isBatchExecuted[messageHash]) revert AlreadyExecuted();
```
or `isMessageExecuted`

We're using 3 different terms to represent the same thing which is confusing. Let's stick to batch?